File: products/amliq/web/src/api/enforcement.ts
import { api } from './client';

export interface EnforcementAction {
  id: string;
  entity_name: string;
  regulator: string;
  action_type: string;
  amount?: number;
  currency?: string;
  date: string;
  description: string;
  url: string;
  jurisdiction: string;
}

interface EnforcementSearchResponse {
  query: string;
  results: EnforcementAction[];
  total: number;
}

interface EnforcementEntityResponse {
  entity_id: string;
  actions: EnforcementAction[];
  total: number;
}

export const enforcementApi = {
  search: (query: string) =>
    api.get<EnforcementSearchResponse>(`/enforcement/search?q=${encodeURIComponent(query)}`),
  getByEntity: (entityId: string) =>
    api.get<EnforcementEntityResponse>(`/enforcement/entity/${entityId}`),
};

File: products/amliq/api/internal/domain/enforcement.go
package domain

import (
	"fmt"
	"time"
)

// EnforcementActionType categorizes regulatory actions.
type EnforcementActionType string

const (
	ActionFine              EnforcementActionType = "fine"
	ActionBan               EnforcementActionType = "ban"
	ActionWarning           EnforcementActionType = "warning"
	ActionLicenseRevocation EnforcementActionType = "license_revocation"
	ActionCeaseDesist       EnforcementActionType = "cease_desist"
)

// EnforcementAction records a regulatory action against an entity.
type EnforcementAction struct {
	ID           string
	EntityName   string
	EntityType   string
	Regulator    string
	ActionType   EnforcementActionType
	Amount       *float64
	Currency     string
	Date         time.Time
	Description  string
	URL          string
	Jurisdiction string
	CreatedAt    time.Time
}

func NewEnforcementAction(
	entityName, regulator string, actionType EnforcementActionType,
	date time.Time, description, url, jurisdiction string,
) (EnforcementAction, error) {
	if entityName == "" || regulator == "" {
		return EnforcementAction{}, fmt.Errorf("entity name and regulator required")
	}
	return EnforcementAction{
		ID:           fmt.Sprintf("enf_%d", time.Now().UnixNano()),
		EntityName:   entityName,
		Regulator:    regulator,
		ActionType:   actionType,
		Date:         date,
		Description:  description,
		URL:          url,
		Jurisdiction: jurisdiction,
		CreatedAt:    time.Now().UTC(),
	}, nil
}
